Spring Boot - Working with third-party API

sshestaka opened the pull request:
- Created API has two methods:
1. The request randomly generates a wiki about one character in the universe the animated series Rick & Morty (according the task)
2. The request takes a string as an argument, and returns a list of all characters whose name contains the search string.

- All data is loaded from the source: "https://rickandmortyapi.com/api/character/"

- Data loading occurs one-time during the launch of the application.

- Liquibase with changelog files have been added.

- MySQL DB is the main data base.

- H2 DB used for tests.

- All requests were documented using Swagger. Swagger is available by path: http://localhost:8080/swagger-ui/index.html#/

sshestaka committed:
Spring Boot - Working with third-party API

- Created API has two methods:
1. The request randomly generates a wiki about one character in the universe the animated series Rick & Morty (according the task)
2. The request takes a string as an argument, and returns a list of all characters whose name contains the search string.

- All data is loaded from the source: "https://rickandmortyapi.com/api/character/"

- Data loading occurs one-time during the launch of the application.

- Liquibase with changelog files have been added.

- MySQL DB is the main data base.

- H2 DB used for tests.

- All requests were documented using Swagger. Swagger is available by path: http://localhost:8080/swagger-ui/index.html#/

olekskov commented on src/main/java/mate/academy/rickandmorty/Application.java:
```diff
@@ -5,8 +5,11 @@
 
 @SpringBootApplication
 public class Application {
+    //private static final CharacterService characterService =
+    //        new CharacterServiceImpl();
```
Please remove commented code

olekskov commented on src/main/java/mate/academy/rickandmorty/Application.java:
```diff
@@ -5,8 +5,11 @@
 
     public static void main(String[] args) {
         SpringApplication.run(Application.class, args);
+        //characterService.saveAll();
```
same

olekskov commented on src/main/java/mate/academy/rickandmorty/controler/CharacterController.java:
```diff
@@ -0,0 +1,54 @@
+        return characterService.searchCharactersByName(searchParametersDto, pageable);
+    }
+
+    @GetMapping("/download/all")
+    @Operation(summary = "Download all characters",
+            description = "Download a list of all available characters from "
+                    + "https://rickandmortyapi.com/api/character/")
+    public List<Character> downloadAll() {
+        return characterService.downloadAllCharacters();
+    }
```
Anyone who finds this endpoint may spam external api on your behalf. Please do not leave such obvious vulnerabilities in code

sshestaka replied:
OK, thanks!

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,98 @@
+    private final CharacterRepository characterRepository;
+    private final CharacterMapper characterMapper;
+
+    @PostConstruct
+    public void init() {
+        downloadAllCharacters();
+    }
+
+    @Override
+    public List<Character> downloadAllCharacters() {
+        HttpClient httpClient = HttpClient.newHttpClient();
+        HttpRequest httpRequest = HttpRequest.newBuilder()
+                .GET()
+                .uri(URI.create(GET_ALL_CHARACTERS_URL))
+                .build();
+        CharacterResponseDataDto dataDto;
+        try {
+            HttpResponse<String> response = httpClient.send(httpRequest,
+                    HttpResponse.BodyHandlers.ofString());
+            dataDto = objectMapper.readValue(response.body(),
+                    CharacterResponseDataDto.class);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        List<CharacterResults> characterResults = dataDto.getResults();
+        List<CharacterDto> characterDtos = characterResults.stream()
+                .map(characterMapper::fromApiToDto)
+                .toList();
+        List<Character> characterList = characterDtos.stream()
+                .map(characterMapper::toModel)
+                .toList();
+        deleteAllCharacters();
+        return characterRepository.saveAll(characterList);
+    }
```
Please extract this part to separate component, it's should not be part of character service

sshestaka replied:
OK, I have moved this logic to the separate class...

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,98 @@
+
+    @Override
+    public CharacterDto findByRandomId() {
+        List<Character> allCharactersFromDb = characterRepository.findAll();
+        int maxIdInDb = allCharactersFromDb.size();
```
Please do not fetch more data from DB than required. 
Ok that this data has few hundreds of rows, what would you do if there were few millions, also make list with all of them? :)
```suggestion
        int maxIdInDb = characterRepository.count();
```

sshestaka replied:
Of corse))

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,98 @@
+                        + generatedLong)));
+    }
+
+    @Override
+    public void deleteAllCharacters() {
+        characterRepository.deleteAll();
+    }
```
Looks strange as public method, it's used only internally. Please remove from interface and inline in implementation

sshestaka committed:
1. Removed commented code
2. Removed URL from API description
3. Data download logic moved to the separate class.
4. Refactored the logic which calculate first and last id in the DB in findByRandomId method. For this purpose I have been added a new method getMaxIdFromCharacters in the CharacterRepository.
5. Refactored access modifier for deleteAllCharacters - to private.

olekskov commented on src/main/java/mate/academy/rickandmorty/controler/CharacterController.java:
```diff
@@ -0,0 +1,53 @@
+        return characterService.searchCharactersByName(searchParametersDto, pageable);
+    }
+
+    @GetMapping("/download/all")
+    @Operation(summary = "Download all characters",
+            description = "Download a list of all available characters")
+    public List<Character> downloadAll() {
+        return characterService.saveAll();
+    }
```
Метод так і не видалено. Навіщо він тобі, якщо ти використовуєш postConstruct? Логіка і так запуститься без ендпоінту взагалі

sshestaka replied:
Я не зрозумів що треба було його видалити... Я навмистно добавив цей метод, щоб коли аплікейшн вже запущено, можна було ще раз перезавантажити контент)) мені здалося що це буде корисно... але зараз видаляю)

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java:
```diff
@@ -0,0 +1,74 @@
+        saveAll();
+    }
+
+    @Override
+    public List<Character> saveAll() {
+        List<CharacterResults> characterResults = charactersClient.downloadAllCharacters();
+        List<Character> characterList = characterResults.stream()
+                .map(characterMapper::fromApiToDto)
+                .map(characterMapper::toModel)
+                .toList();
+        deleteAllCharacters();
+        return characterRepository.saveAll(characterList);
+    }
```
Наявність методу saveAll() без аргументів дуже спантеличує. Загалом сервіс собі працює з базою і дтошками, а тут чогось йому треба йти кудись на стороннє апі і щось там витягувати
```suggestion
```

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharacterService.java:
```diff
@@ -0,0 +1,18 @@
+import org.springframework.data.domain.Pageable;
+
+public interface CharacterService {
+    List<Character> saveAll();
```
```suggestion
```

olekskov commented on src/main/java/mate/academy/rickandmorty/service/CharactersClient.java:
```diff
@@ -0,0 +1,40 @@
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class CharactersClient {
+    private final ObjectMapper objectMapper;
+
+    @Value("${mate.academy.get.all.characters.url}")
+    private String getAllCharactersUrl;
+
+    public List<CharacterResults> downloadAllCharacters() {
+        HttpClient httpClient = HttpClient.newHttpClient();
+        HttpRequest httpRequest = HttpRequest.newBuilder()
+                .GET()
+                .uri(URI.create(getAllCharactersUrl))
+                .build();
+        CharacterResponseDataDto dataDto;
+        try {
+            HttpResponse<String> response = httpClient.send(httpRequest,
+                    HttpResponse.BodyHandlers.ofString());
+            dataDto = objectMapper.readValue(response.body(), CharacterResponseDataDto.class);
+        } catch (IOException | InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+        return dataDto.getResults();
+    }
+}
```
Весь код наповнення бази логічніше виглядає тут
```suggestion
@Component
@RequiredArgsConstructor
public class CharactersClient {
    private final ObjectMapper objectMapper;
    
    //add repository

    @Value("${mate.academy.get.all.characters.url}")
    private String getAllCharactersUrl;

    @PostConstruct
    private List<CharacterResults> downloadAllCharacters() {
        HttpClient httpClient = HttpClient.newHttpClient();
        HttpRequest httpRequest = HttpRequest.newBuilder()
                .GET()
                .uri(URI.create(getAllCharactersUrl))
                .build();
        CharacterResponseDataDto dataDto;
        try {
            HttpResponse<String> response = httpClient.send(httpRequest,
                    HttpResponse.BodyHandlers.ofString());
            dataDto = objectMapper.readValue(response.body(), CharacterResponseDataDto.class);
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        List<Character> characterList = dataDto.getResults().stream()
                .map(characterMapper::fromApiToDto)
                .map(characterMapper::toModel)
                .toList();
        characterRepository.deleteAll();
        characterRepository.saveAll(characterList);
    }
}
```

sshestaka replied:
Я зрозумів) Взагалі видаляю метод save в CharacterService і переношу всю логіку по збереженню даних в CharactersClient!

sshestaka committed:
1. Method downloadAll removed from CharacterController.
2. Method saveAll and all logic connected with save a new data removed from CharacterService.
3. Class CharactersClient have been refactored: added logic for work with the DB and deleteAll / saveAll methods.

Ivan95kos reviewed (approved):
Nice
